### ISSUE TITLE
Split NAudio into multiple packages

### ISSUE THREAD

**Issue by jwosty** [naudio/NAudio#591]
After reading #184 I think it is time to take the .Net Standard spirit to NAudio, and split it into multiple packages that each target different platforms. The first step would be to identify the completely platform-independent interfaces and classes (for example IWavePlayer and WdlResamplingSampleProvider) and refactor them out into a new package, perhaps NAudio.Core, keeping the same namespaces for backward-compatibility. The next step would be to move the various platform/api-specific classes and interfaces into their own separate packages each; for example putting all the Asio stuff into a new package called NAudio.Asio.

This is a huge step forward for opening up the possibility for new platform implementations. For example, a motivated person could create macOS implementations of the NAudio API, and call it NAudio.AppleCoreAudio in the future. Here's a quick-n-dirty outline of the possible new package structure:

* NAudio.Core (.Net Standard 2.0)
  * Shared interfaces and functionality like IWavePlayer and WdlResamplingSampleProvider
* NAudio.Asio (.Net Standard) (Windows-only)
* NAudio.WaveOut (.Net Framework 3, .Net Core 3.0) (Windows-only)
* NAudio.WindowsCoreAudio (.Net Standard or whatever) (Windows-only)
* _NAudio.AppleCoreAudio (Xamarin.Mac)_
  * in the future, someone could certainly create this as a third-party library, or even submit a PR to the official one to implement this
  * macOS codecs and audio playback/recording APIs, probably based on AudioQueues

_(I could also imagine breaking things down further, like splitting playback and recording into different packages, though this may be overkill)_

Finally, we would also have to make sure to create a new NAudio package that includes no code but simply includes all the other packages directly, again for backward compatibility, as not to break anything that directly reference the "NAudio" package. Probably also mark it as deprecated.

I'm willing to take this on, as it is mostly dumb refactoring; although if someone beats me too this I won't be too upset :)

**Comment by markheath** [comment]
This is something I've given quite a lot of thought to, and if there were to be a NAudio version 2 I'd definitely break it up into smaller chunks.

My ideas are similar with an NAudio.Core package that contains all the interfaces, plus all pure cross-platform .NET standard code (e.g. things like WaveFileReader and pure C# codec or effect implementations). 

Then there would be a series of interop packages for the various APIs - e.g. NAudio.Asio.Interop, NAudio.Mme.Interop (waveIn, waveOut), NAudio.Wasapi.Interop, etc. These would ideally not depend on NAudio.Core, so if someone wanted to use those interop wrappers in another project but not use the NAudio abstractions they would be free to do so (this would be particularly useful for ASIO which isn't a great fit for the existing NAudio interfaces). For some of these packages we could even switch to using other open source packages such as SharpDX's mediafoundation which has some advantages over NAudio's interop (although it's sadly not under development any more, so might no longer be an option)

Finally, various packages containing implementations of NAudio interfaces and base classes that depend on both NAudio.Core and one (or more) of the interop packages. So for example, WasapiOut could be in an NAudio.WindowsCoreAudio package, and MediaFoundationReader in a NAudio.MediaFoundation package. I haven't quite decided what the best way to slice this up, but something similar to your suggested breakdown would make sense. And as you point out would also make it obvious how to produce an extension for other operating systems (although it is possible today thanks to the NetStandard target)

One of the trickiest parts of splitting everything out in this way is the fact that I use `WaveFormat` extensively in NAudio, which is a struct based pretty closely on `WAVEFORMATEX`. The difficulty here is that only some of the audio APIs are based on `WAVEFORMATEX` (waveIn, waveOut, acm ... etc). WASAPI, ASIO etc have their own way of defining wave formats. I've not quite decided what the best solution to this is, but probably, the "interop" packages should always use a struct matching what the underlying API uses, and the API specific implementation packages would know how to convert from those structs into an NAudio specific `WaveFormat` class, which may for simplicity remain very close to what the existing one is. 

As I said, I was thinking of saving this for NAudio 2, as I am very reluctant to introduce any kind of breaking change. Another technical challenge is simply how to have one repository that builds multiple NuGet packages that depend on each other. I'd want to be able to easily debug into all the code (so use project references for debugging), but in the release build make everything a NuGet package with the dependencies wired up. I'm sure this challenge has been solved by someone, but I've not found a good example that I want to copy yet.

Another goal of NAudio 2 is to enable me to rework some of the interfaces. Particularly changing to a `Span<T>` based WaveProvider, and cleaning up a few other interfaces as well.

In short, yes I do see the future of NAudio as being a collection of smaller packages, but would like to get the design right before pulling the trigger on it.

**Comment by jwosty** [comment]
The simple act of dividing up the classes that currently exist among some new packages can be done in a completely backwards-compatible way, as long as you don’t change any of the actual API (the API changes that you mentioned could be a separate concern and feature PR). I’ve started experimenting with this in #593.

> Another technical challenge is simply how to have one repository that builds multiple NuGet packages that depend on each other.

`nuget pack` actually already handles this. If you have multiple sdk-style projects that reference each other, the pack command (also `dotnet pack`) will make budget packages out of each of them and then generate the appropriate PackageReference command*, no nuspec necessary.

*There’s just one problem with this process. Automatic packing will always generate a >= constraint on the other project libraries which is usually not what you want — usually you want a == constraint. See NuGet/Home/issues/5540. I’ve worked around this exact problem in my own library here by hacking the .nupkg after it’s built until nuget gets fixed: jwosty/Interstellar/issues/3 (fix is at [16533d33d](jwosty/Interstellar/commit/16533d33d983d71303e79bf9ec7d426c95f49898))

**Comment by jwosty** [comment]
also, How can I help with a design for this and/or NAudio 2? Should we draft up a technical document (a la RFCs) specifying the design, which can be used for collaboration? An issue thread alone could be a little hard to keep track of. I have some relevant knowledge of the macOS audio APIs that could inform new designs. If we take some breaking changes for a 2.0 version I think it should be in service of cross platform concerns, because I believe that can be a strong foundation for NAudio 2.0.

**Comment by markheath** [comment]
Wow I didn't know about the `nuget pack` feature. I'd been wondering for ages on how to do this.

And yes, it would be good to discuss the breakdown. Maybe you can give me a bit more detail on how you see the existing bits of NAudio being broken into sub-packages (i.e. not thinking NAudio 2 just yet). I think there will be a few interesting choices as to exactly what gets bundled together and they will be tricky to name well. If we can produce a v1.11 in a completely backwards compatible way that would be great.

**Comment by jwosty** [comment]
As a user of this library, a core package + a package for each API implementation would be useful: that way you could include only the platforms you plan on using. Beyond that I'm not sure as I'm not as familiar with some of the other aspects of the library (for example I see there is MIDI stuff). Perhaps something like this?

* NAudio.Core - core interfaces, any functionality written in pure C# (including WdlResampler, all the C# codec implementations, all the C# WaveProviders, etc)
* NAudio.Asio - ASIO implementations of IWavePlayer, IWaveIn, etc, as well as their direct dependencies
* NAudio.Wasapi - Wasapi implementations
* NAudio.WaveOut - WaveIn and WaveOut implementations - not a good name, but I don't personally know what to call it
* NAudio.DirectSound - DirectSound implementations
* NAudio.Codecs.Windows - Windows-only codecs, like Mp3FileReader

This leaves the option for adding more of these packages down the line, like an implementation for macos and/or linux (I would consider a macos package much higher priority)

**Comment by DennisAkaXIT** [comment]
Hey all,

I am using the NAudio Midi package since a while now and found this post after running into trouble by creating some of my platform independent Midi / Multimedia parts.

My basic structure is something like this:

.NET platform independent -> .NET Windows -> .NET WPF -> .NET Application

Having all MIDI related parts in NAudio.Midi makes it impossible to create some additional platform independent parts like a MidiPattern class which depends on MidiEventCollections, or even just opening a MidiFile, without compiling our own versions of NAudio.

Therefore, I have a few additional suggestions for this post.

- Move all basic Midi XYZ Events and MidiFile to NAudio.Core.Midi (as far as I could see, there is no real dependecy to anything), except of MidiInCapabilities.cs, MidiIn.cs, MidiOutCapabilities.cs, MidiOut.cs.
- Move the MidiInterop.cs and MidiOutTechnology.cs to NAudio.WinMM.Midi, to keep all WinMM parts at one place
- Having less files in this folder (at the end there would be just a few) could inspire others to port the MidiIn and MidiOut classes to other platforms, because of the better overview.

Maybe additional:
-  Create an abstract base class (or an interface) for MidiIn and MidiOut, which leaves the opportunity to port it for a different platform, but having a basic structure to follow. (I think an interface would be the better idea. For example IMidiOut = Platform.GetMidiOut(index))
- Creating an additional Package named NAudio.WinMidi instead of NAudio.Midi (I know this would be a problem for many users, but this would keep it cleaner if "some day" multiple platform implementations exist. (I want to do that in my program/game in the future)

For now, I just want to know if this (or parts of my suggestions) would be an option for you. If so, I could test all of it (or parts) with the actual master and if it works, I could send a pull reqest.

Please let me know what you think.

Regards,
Dennis

**Comment by hansmbakker** [comment]
I'm just getting started with NAudio, but it seems like a very useful and nice library!

I see work on this issue is being done, but is there any status overview in a PR of what is done and what still needs to be done? Or a roadmap?

It is good that the packages are being restructured but I feel the package tree and dependency tree still have some room for improvement:
- for example, the `WpfDemo` references the `NAudio.WinForms` code, even though it is a WPF app?
  https://github.com/naudio/NAudio/blob/03b46c70049628ba2237ff9ddce0337143ebcc76/NAudioWpfDemo/NAudioWpfDemo.csproj#L11
- `NAudio.Extras` depends on `NAudio` which pulls in everything (even when you don't make a WinForms app, for example), on top of the specific packages it referenced.
  https://github.com/naudio/NAudio/blob/03b46c70049628ba2237ff9ddce0337143ebcc76/NAudio.Extras/NAudio.Extras.csproj#L19-L21
- the `NAudio` package is kept for the purpose of an umbrella package, but it seems to contain two leftover classes. Maybe those convenience classes could go somewhere else?
  - there is the `Mp3FileReader` code which seems like it does not have platform-specific dependencies?
  - The `AudioFileReader` seems to depend only on NAudio.Core and NAudio.Wasapi (for the fallback).

**Comment by markheath** [comment]
It would be nice to tidy up the package structure a bit further, but probably in the short term things will stay where they are. The WPF thing is because the window callback mode for WaveOut uses WinForms.

**Comment by Swimburger** [comment]
Does the `WaveFormatConversionStream` API depend on any Windows APIs? 
If not, it would be great if the conversion APIs are available cross-platform, which I'd assume would go under Core?

I'm trying to convert a byte array from PCM S16LE to MuLaw on MacOS, but I want this application to work everywhere.

**Comment by markheath** [comment]
Yes, WaveFormatConversionStream is a wrapper around the Windows ACM APIs. If you just need mu-law and a-law conversion, there are fully managed conversion algorithms for those in NAudio that you can use instead.

**Comment by markheath** [comment]
Going to close this old issue - NAudio is now broken up into several packages, and with NAudio 3 there are some new ones such as NAudio.Dmo (which holds the legacy DMO and DirectSound capabilities), NAudio.WinMM (for the legacy windows multimedia APIs), and NAudio.Wasapi (which bundles WASAPI with Media Foundation - the modern Windows Audio stack), plus several others such as NAudio.Asio, NAudio.Alsa etc.

I've made the split much cleaner now as well - everything in NAudio.Core is truly cross-platform, and then you can add in platform or API specific packages for the extra things you need
